hie-plugin-api/Haskell/Ide/Engine/Cradle.hs

```diff
@@ -534,25 +538,78 @@ cabalHelperCradle file = do
-      warningm
-        $  "The Exception was thrown in the context of finding"
-        ++ " a component for \""
-        ++ fp
-        ++ "\" in the unit: "
-        ++ show unit
-      getComponent env units fp
-    Right ui -> do
-      let components = M.elems (uiComponents ui)
-      debugm $ "Unit Info: " ++ show ui
-      case find (fp `partOfComponent`) components of
-        Nothing -> getComponent env units fp
-        comp    -> return comp
+  :: forall pt. QueryEnv pt -> [Unit pt] -> FilePath -> IO (Either [String] ChComponentInfo)
+getComponent env unitCandidates fp = getComponent' [] [] unitCandidates >>=
+    \case
+      (tried, failed, Nothing) -> return (Left $ buildErrorMsg tried failed)
+      (_, _, Just comp) -> return (Right comp)
+  where
+    getComponent' :: [UnitInfo] -> [(Unit pt, IOException)] -> [Unit pt] -> IO ([UnitInfo], [(Unit pt, IOException)], Maybe ChComponentInfo)
+    getComponent' triedUnits failedUnits [] = return (triedUnits, failedUnits, Nothing)
+    getComponent' triedUnits failedUnits (unit : units) =
+      try (runQuery (unitInfo unit) env) >>= \case
+        Left (e :: IOException) -> do
+          warningm $ "Catching and swallowing an IOException: " ++ show e
+          warningm
+            $  "The Exception was thrown in the context of finding"
+            ++ " a component for \""
+            ++ fp
+            ++ "\" in the unit: "
+            ++ show unit
+          getComponent' triedUnits ((unit, e):failedUnits) units
+        Right ui -> do
+          let components = Map.elems (uiComponents ui)
+          debugm $ "Unit Info: " ++ show ui
+          case find (fp `partOfComponent`) components of
+            Nothing -> getComponent' (ui:triedUnits) failedUnits units
+            comp    -> return (triedUnits, failedUnits, comp)
+
+    buildErrorMsg :: [UnitInfo] -> [(Unit pt, IOException)] -> [String]
+    buildErrorMsg triedUnits failedUnits =
+            [ "Could not obtain flags for: \"" ++ fp ++ "\"."
+            , ""
+            ]
+            ++ concat
+              [
+                [ "This Module was not part of any component we are aware of."
+                , ""
+                ]
+                ++ concatMap ppShowUnitInfo triedUnits
+                ++ [ ""
+                   , ""
+                   , "If you dont know how to expose a module, take a look at:"
+                   , "https://www.haskell.org/cabal/users-guide/developing-packages.html"
+                   , ""
+                   ]
+              | not (null triedUnits)
+              ]
+            ++ concat
+              [
+                [ "We could not build all components."
+                , "If one of these components exposes this Module, make sure they compile."
+                , "You can try to invoke the commands yourself."
+                , "The following commands failed:"
+                ]
+                ++ concatMap (ppShowIOException . snd) failedUnits
```

Comment:
This is so much more user-friendly 👍

Reply:
It is still horrible, but we discussed on irc how to improve cabal-helper to have good error messages!